File: integrador-python/modulo3_integrador/infrastructure/redis_listener.py
import redis
import json
import threading
from typing import Callable, Any


class RedisListener:
    def __init__(self, redis_url: str = "redis://localhost:6379", channel: str = "crud-channel"):
        self.redis_client = redis.from_url(redis_url, decode_responses=True)
        self.pubsub = self.redis_client.pubsub()
        self.channel = channel
        self.message_handler: Callable[[str, str], None] = None
        self.is_listening = False
        self.listener_thread = None
    
    def set_message_handler(self, handler: Callable[[str, str], None]):
        """Define o handler para processar mensagens recebidas"""
        self.message_handler = handler
    
    def start(self):
        """Inicia a escuta do canal Redis"""
        if self.is_listening:
            print("⚠️ Listener já está ativo")
            return
        
        if not self.message_handler:
            raise ValueError("Message handler deve ser definido antes de iniciar")
        
        self.pubsub.subscribe(self.channel)
        self.is_listening = True
        
        print(f"🟢 Escutando canal Redis: {self.channel}")
        
        # Inicia thread para escutar mensagens
        self.listener_thread = threading.Thread(target=self._listen_loop, daemon=True)
        self.listener_thread.start()
    
    def _listen_loop(self):
        """Loop principal para escutar mensagens"""
        try:
            for message in self.pubsub.listen():
                if not self.is_listening:
                    break
                
                if message['type'] == 'message':
                    channel = message['channel']
                    data = message['data']
                    
                    print(f"🔔 [{channel}] {data}")
                    
                    # Chama o handler definido
                    if self.message_handler:
                        try:
                            self.message_handler(channel, data)
                        except Exception as e:
                            print(f"❌ Erro ao processar mensagem: {e}")
        except Exception as e:
            print(f"❌ Erro no loop de escuta: {e}")
        finally:
            self.is_listening = False
    
    def stop(self):
        """Para a escuta do canal Redis"""
        if not self.is_listening:
            return
        
        self.is_listening = False
        self.pubsub.unsubscribe(self.channel)
        self.pubsub.close()
        
        if self.listener_thread and self.listener_thread.is_alive():
            self.listener_thread.join(timeout=5)
        
        print("🛑 Listener Redis parado")
    
    def close(self):
        """Fecha a conexão Redis"""
        self.stop()
        self.redis_client.close()

File: modulo3_integrador_apache_camel/src/main/java/br/ufg/inf/integrador/redis/RedisListener.java
package br.ufg.inf.integrador.redis;

import io.lettuce.core.RedisClient;
import io.lettuce.core.pubsub.RedisPubSubAdapter;
import io.lettuce.core.pubsub.StatefulRedisPubSubConnection;
import org.apache.camel.ProducerTemplate;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Listener Redis que consome eventos CRUD e os envia para o Apache Camel
 * 
 * Implementa o padrão Publisher-Subscriber Channel:
 * - Subscreve ao canal 'crud-channel' do Redis
 * - Recebe eventos de operações CRUD dos sistemas SGA/SB
 * - Encaminha mensagens para o Apache Camel via ProducerTemplate
 */
public class RedisListener {
    
    private static final Logger logger = LoggerFactory.getLogger(RedisListener.class);
    private static final String REDIS_URL = "redis://localhost:6379";
    private static final String CRUD_CHANNEL = "crud-channel";
    private static final String CAMEL_ENDPOINT = "direct:crud";
    
    private final ProducerTemplate producer;
    private RedisClient redisClient;
    private StatefulRedisPubSubConnection<String, String> connection;
    
    public RedisListener(ProducerTemplate producer) {
        this.producer = producer;
    }
    
    /**
     * Inicia a escuta do canal Redis
     */
    public void start() {
        try {
            logger.info("🔗 Conectando ao Redis: {}", REDIS_URL);
            
            // Criar cliente Redis
            redisClient = RedisClient.create(REDIS_URL);
            connection = redisClient.connectPubSub();
            
            // Configurar listener para mensagens
            connection.addListener(new RedisPubSubAdapter<String, String>() {
                @Override
                public void message(String channel, String message) {
                    logger.info("🔔 [{}] {}", channel, message);
                    
                    try {
                        // Enviar mensagem para o Apache Camel
                        producer.sendBody(CAMEL_ENDPOINT, message);
                        logger.debug("📤 Mensagem enviada para Camel: {}", CAMEL_ENDPOINT);
                        
                    } catch (Exception e) {
                        logger.error("❌ Erro ao enviar mensagem para Camel: {}", e.getMessage(), e);
                    }
                }
                
                @Override
                public void subscribed(String channel, long count) {
                    logger.info("✅ Subscrito ao canal: {} (total: {})", channel, count);
                }
                
                @Override
                public void unsubscribed(String channel, long count) {
                    logger.info("🔕 Desconectado do canal: {} (restantes: {})", channel, count);
                }
            });
            
            // Subscrever ao canal CRUD
            connection.sync().subscribe(CRUD_CHANNEL);
            logger.info("🟢 Escutando canal Redis: {}", CRUD_CHANNEL);
            
        } catch (Exception e) {
            logger.error("❌ Erro ao iniciar Redis listener: {}", e.getMessage(), e);
            throw new RuntimeException("Falha ao conectar ao Redis", e);
        }
    }
    
    /**
     * Para a escuta e fecha conexões
     */
    public void stop() {
        try {
            if (connection != null) {
                connection.sync().unsubscribe(CRUD_CHANNEL);
                connection.close();
                logger.info("🔕 Conexão Redis fechada");
            }
            
            if (redisClient != null) {
                redisClient.shutdown();
                logger.info("🛑 Cliente Redis finalizado");
            }
            
        } catch (Exception e) {
            logger.error("⚠️ Erro ao parar Redis listener: {}", e.getMessage(), e);
        }
    }
}